pkg/reconciler/ingress/ingress.go

```diff
@@ -0,0 +1,279 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ingress
+
+import (
+	"context"
+	"fmt"
+	"hash/fnv"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+const (
+	clusterLabel = "kcp.dev/cluster"
+	ownedByLabel = "kcp.dev/owned-by"
+)
+
+// reconcile is triggered on every change to an ingress resource, or it's associated services (by tracker).
+func (c *Controller) reconcile(ctx context.Context, ingress *networkingv1.Ingress) error {
+	klog.Infof("reconciling Ingress %q", ingress.Name)
+
+	if ingress.Labels == nil || ingress.Labels[clusterLabel] == "" {
+		// This is a root Ingress; get its leafs.
+		sel, err := labels.Parse(fmt.Sprintf("%s=%s", ownedByLabel, ingress.Name))
+		if err != nil {
+			return err
```

Comment:
Log & return nil - this will be a permanent failure and will go into rate limited infinite backoff if you return the error.

Please add a TODO to set a user-visible condition describing the error